Feature/migrate merchandise return view 1

JevgenijVisockij opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR migrates "merchandise return" view/edit BO page to symfony. This PR migrates only view of merchandise return **without product list or label printing** as planned in roadmap on https://github.com/PrestaShop/PrestaShop/issues/24421.
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | One item of https://github.com/PrestaShop/PrestaShop/issues/24421
| How to test?  | See below

Reopening of https://github.com/PrestaShop/PrestaShop/pull/24154
## How to test

1. Enable merchandise returns option
2. Create some merchandise returns.
3. Go to `admin-dev/index.php/sell/customer-service/merchandise-return/` which is the list. Click on any of existing returns to open the page.
4. In this page, you should be able to change the 'order return state' and save it successfully. The design is not very good because the page is not complete.

⚠️ **PDF printing and product list are not included in this PR.** They will be implemented later. Only changing the state is implemented here. The design is not very good because the page is not complete. It will look so much better when next PRs are submitted and merged.

## About class and namespace naming

All new classes and namespaces in this PR are `OrderReturn` instead of `MerchandiseReturn` because that's the direction we decided to go with.

Everything codewise in legacy was called OrderReturn as well so we want to maintain codebase consistency. So OrderReturn in code = MerchandiseReturn in interface.

There will be a separate PR to change all existing code(mostly related to Merchandise Return grid) to reflect that (see roadmap https://github.com/PrestaShop/PrestaShop/issues/24421).


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/19420)
<!-- Reviewable:end -->

JevgenijVisockij committed:
Fixed integration test

JevgenijVisockij committed:
Apply suggestions from code review

Small fixes from code review

Co-authored-by: GoT <PierreRambaud@users.noreply.github.com>

JevgenijVisockij committed:
Apply suggestions from code review

Wording domain changes

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Added cache clearing and removed unneeded step

JevgenijVisockij committed:
Update src/Adapter/OrderReturn/QueryHandler/GetOrderReturnForEditingHandler.php

Removed final from handler

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Apply suggestions from code review

Removed some uneeded final classes, also some uneeded spaces

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
ADded unit tests for merchandise return

JevgenijVisockij committed:
Fixed issue with symbols < not showing as escaped values in order return customer explanation.

JevgenijVisockij committed:
Removed no longer needed javascript

JevgenijVisockij committed:
Fixed issue with wrong exception thrown when order return is not found

JevgenijVisockij committed:
Changed order state update logic to use domain services

JevgenijVisockij committed:
Fixed error codes, added error code to OrderReturnId exception. Added warning to OrderReturn create method

JevgenijVisockij committed:
Update src/PrestaShopBundle/Controller/Admin/Sell/CustomerService/MerchandiseReturnController.php

Changed from checknig for id instead to check for isValid and is submitted when saving order return

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/config/services/core/form/form_handler.yml

Using Symfony 4.4 syntax for factory

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

JevgenijVisockij committed:
Update tests/Integration/Behaviour/Features/Context/Domain/OrderReturnFeatureContext.php

Using getter rather then static function

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

JevgenijVisockij committed:
Update tests/Integration/Behaviour/Features/Scenario/OrderReturn/order_return_management.feature

Adding line break for clarity

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

JevgenijVisockij committed:
Refactored test to use order return for editing

JevgenijVisockij committed:
Renamed order return state back as it shouldn't part of this PR

JevgenijVisockij committed:
Fixed factory definition in services and removed no longer used path in forms save

JevgenijVisockij committed:
Apply suggestions from code review

Suggestions for exception comments

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Apply suggestions from code review

Corrected service names

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Added test for order return data provider

JevgenijVisockij committed:
Apply suggestions from code review

Made OrderReturn::get smaller

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Fixed order return context namespace

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Fixed test

JevgenijVisockij committed:
Added check for duplicate order names

JevgenijVisockij committed:
Adding an if for PHP Stan to shut it

JevgenijVisockij committed:
Fixing tests

JevgenijVisockij committed:
Adding ID in front of order return state to avoid duplicate order names issues

JevgenijVisockij committed:
Using sprintf

JevgenijVisockij committed:
Adding docblock

JevgenijVisockij committed:
Fixes to yml files

JevgenijVisockij committed:
Adding missing return type

JevgenijVisockij committed:
Adding missing return type

JevgenijVisockij committed:
Update src/Adapter/OrderReturn/Repository/OrderReturnRepository.php

IMO makes sense.

Co-authored-by: Mathieu Ferment <mathieu.ferment@prestashop.com>

JevgenijVisockij committed:
Added missing )

matks committed:
Apply suggestions from code review

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Migrating base for merchandise return view page

# Conflicts:
#	src/PrestaShopBundle/Resources/config/routing/admin/sell/customer_service/merchandise_return.yml
#	src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml

JevgenijVisockij committed:
Moved order return test after order

# Conflicts:
#	tests/Integration/Behaviour/behat.yml

# Conflicts:
#	tests/Integration/Behaviour/behat.yml

JevgenijVisockij committed:
Changed order return update to use customer and order repositories instead of using legacy objects outright.

Renamed EditableOrderReturn and EditableOrderReturnState into OrderReturnForEditing and OrderReturnStateForEditing

# Conflicts:
#	src/PrestaShopBundle/Resources/config/services/adapter/customer.yml
#	templates/bundles/PrestaShopBundle/Admin/Configure/ShopParameters/OrderReturnStates/edit.html.twig

JevgenijVisockij committed:
Update src/PrestaShopBundle/Resources/config/routing/admin/sell/customer_service/merchandise_return.yml

Using full notation to avoid compatability problems

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>
# Conflicts:
#	src/PrestaShopBundle/Resources/config/routing/admin/sell/customer_service/merchandise_return.yml

JevgenijVisockij committed:
Removed some uneeded things like sprintf or DataProvider

# Conflicts:
#	src/PrestaShopBundle/Resources/config/routing/admin/sell/customer_service/merchandise_return.yml

JevgenijVisockij committed:
Renamed order return state back as it shouldn't part of this PR

# Conflicts:
#	templates/bundles/PrestaShopBundle/Admin/Configure/ShopParameters/OrderReturnStates/edit.html.twig

JevgenijVisockij committed:
Moved order return and order return state creation to a different context

# Conflicts:
#	tests/Integration/Behaviour/behat.yml

JevgenijVisockij committed:
CS Fixer fixes

# Conflicts:
#	src/Adapter/Customer/Repository/CustomerRepository.php

JevgenijVisockij committed:
Fixed copyright

# Conflicts:
#	src/Adapter/Customer/Repository/CustomerRepository.php

JevgenijVisockij committed:
Apply suggestions from code review

Minor fixes: return types and docs

Co-authored-by: Progi1984 <progi1984@gmail.com>
# Conflicts:
#	src/Adapter/Customer/Repository/CustomerRepository.php

Amoifr commented on src/Adapter/Order/Repository/OrderRepository.php:
```diff
@@ -0,0 +1,73 @@
+    public function get(OrderId $orderId): Order
+    {
+        try {
+            $order = new Order($orderId->getValue());
```
May be I missed someting, but the comment announce that this method is about getting legacy Order but it just make a new Order....

zuk3975 replied:
when you provide id, legacy obj model is loaded from db.

Amoifr commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -0,0 +1,84 @@
+    {
+        /* @var OrderReturn $orderReturn */
+        $orderReturn = $this->getObjectModel($orderReturnId->getValue(), OrderReturn::class, OrderReturnNotFoundException::class);
+        if (!$orderReturn instanceof OrderReturn) {
```
just a detail .... but when I read this line. I was not sure that he "!" was interpreted after the rest of the expression. Could you make an explicit parenthesis or inverse the condition and return if true. More readable.
```suggestion
        if (!($orderReturn instanceof OrderReturn)) {
```

Amoifr commented on src/Adapter/OrderReturn/Validator/OrderReturnValidator.php:
```diff
@@ -0,0 +1,70 @@
+     *
+     * @throws CoreException
+     */
+    private function validateOrderReturnProperty(OrderReturn $product, string $propertyName, int $errorCode = 0): void
```
$orderReturn or $product ?

Amoifr commented on src/Core/Form/IdentifiableObject/DataProvider/OrderReturnFormDataProvider.php:
```diff
@@ -0,0 +1,116 @@
+     */
+    public function __construct(
+        CommandBusInterface $queryBus,
+        Router $router,
```
Router or RouterInterface .... I don't know.

JevgenijVisockij replied:
Yeah I guess router interface makes more sense

Amoifr commented on src/Core/Form/IdentifiableObject/DataProvider/OrderReturnFormDataProvider.php:
```diff
@@ -0,0 +1,116 @@
+        return [];
+    }
+
+    private function buildOrderReturnInformation(OrderReturnForEditing $orderReturnForEditing)
```
```suggestion
    private function buildOrderReturnInformation(OrderReturnForEditing $orderReturnForEditing): string
```

JevgenijVisockij committed:
Removing uneeded form types

JevgenijVisockij committed:
Update src/Adapter/OrderReturn/Repository/OrderReturnRepository.php

Adding extra ( just to make condition more clear

Co-authored-by: Amoifr <31698966+Amoifr@users.noreply.github.com>

JevgenijVisockij committed:
Update src/Core/Form/IdentifiableObject/DataProvider/OrderReturnFormDataProvider.php

Adding return type

Co-authored-by: Amoifr <31698966+Amoifr@users.noreply.github.com>

JevgenijVisockij committed:
Fixed $product variable

JevgenijVisockij committed:
Changed to use routerinterface

JevgenijVisockij committed:
Refactored repositories to work properly added integration tests for them.

JevgenijVisockij committed:
Fixed order return saving not working

JevgenijVisockij committed:
CS fixer fixes

zuk3975 commented on src/Adapter/Form/ChoiceProvider/OrderReturnStateChoiceProvider.php:
```diff
@@ -0,0 +1,69 @@
+        $orderStates = OrderReturnState::getOrderReturnStates($this->contextLangId);
+
+        foreach ($orderStates as $orderState) {
+            $choices[sprintf('%s - %s', $orderState['id_order_return_state'], $orderState['name'])]
+                = (int) $orderState['id_order_return_state'];
```
Maybe it is worth introducing a var for this case, it is hard to read

```suggestion
            $indexName = sprintf('%s - %s', $orderState['id_order_return_state'], $orderState['name']);
            $choices[$indexName] = (int) $orderState['id_order_return_state'];
```

margud replied:
+, hard to understand what you want to do here, if you create var with name that explains what it is then it would take less time to understand what you want to do here.

zuk3975 commented on src/Adapter/OrderReturn/CommandHandler/UpdateOrderReturnStateHandler.php:
```diff
@@ -0,0 +1,101 @@
+    public function __construct(
+        OrderReturnRepository $orderReturnRepository,
+        OrderReturnStateRepository $orderReturnStateRepository,
+        OrderReturnValidator $validator
```
```suggestion
        OrderReturnValidator $orderReturnValidator
```

zuk3975 commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -0,0 +1,84 @@
+     */
+    private $validator;
+
+    public function __construct(OrderReturnValidator $validator)
```
```suggestion
    public function __construct(OrderReturnValidator $orderReturnValidator)
```

zuk3975 commented on src/Core/Form/IdentifiableObject/DataProvider/OrderReturnFormDataProvider.php:
```diff
@@ -0,0 +1,116 @@
+        return [];
+    }
+
+    private function buildOrderReturnInformation(OrderReturnForEditing $orderReturnForEditing): string
```
missing docblock

zuk3975 reviewed (commented):
just some minor suggestions :+1:

JevgenijVisockij committed:
Changed validator to orderReturnValidator

JevgenijVisockij committed:
Adding missing doc block

JevgenijVisockij commented:
> just some minor suggestions +1

Added them

orsaz commented on src/Adapter/Form/ChoiceProvider/OrderReturnStateChoiceProvider.php:
```diff
@@ -0,0 +1,69 @@
+    private $contextLangId;
+
+    /**
+     * @param int $contextLangId
```
do we need to add typehint here if we already do it in construct itself?

JevgenijVisockij replied:
Well that's how it's done in the rest of PS so don't really want to deviate from that.

jolelievre replied:
We don't have a strict policy on this, basically we force the PHPDoc for methods with parameters with missing type, when it's fully typed it becomes optional but it's not forbidden either 🤷‍♂️

JevgenijVisockij committed:
Update src/Adapter/Form/ChoiceProvider/OrderReturnStateChoiceProvider.php

Adding variable for clearer reading

Co-authored-by: Julius Žukauskas <31609858+zuk3975@users.noreply.github.com>

JevgenijVisockij committed:
Adding missing doc block

FabienPapet commented on src/PrestaShopBundle/Form/Admin/Sell/CustomerService/MerchandiseReturn/OrderReturnType.php:
```diff
@@ -0,0 +1,110 @@
+            ])
+            ->add('order_return_state', ChoiceType::class, [
+                'required' => true,
+                'choices' => $this->stateChoices,
```
I think this should be an option to allow reusability

FabienPapet commented on src/Adapter/Order/Repository/OrderRepository.php:
```diff
@@ -0,0 +1,73 @@
+use PrestaShop\PrestaShop\Core\Repository\AbstractObjectModelRepository;
+use PrestaShopException;
+
+class OrderRepository extends AbstractObjectModelRepository
```
Maybe create an interface for this that can be used for with the new core services. So we just have to replace them we we migrate

JevgenijVisockij replied:
Good idea regarding interfaces, will add them.

jolelievre replied:
As discussed these days, I don't think we should introduce interfaces too soon. The interfaces we are aiming are generic ones that should be independent from the ObjectModel and based on generic DTOs/interfaces That's not the case here because the ObjectModelRepositories are not meant to be generic ones that will stay in the future. They are mainly helpers services to fetch object models properly with proper exceptions and they avoid duplicating the same code in all the handlers.

That's why they should remain adapter services and don't need an interface, since there is nothing we'll be able to substitute for them.

FabienPapet commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -0,0 +1,87 @@
+use PrestaShop\PrestaShop\Core\Exception\CoreException;
+use PrestaShop\PrestaShop\Core\Repository\AbstractObjectModelRepository;
+
+class OrderReturnRepository extends AbstractObjectModelRepository
```
Should also have an interface

zuk3975 commented on admin-dev/themes/new-theme/scss/components/_form.scss:
```diff
@@ -226,3 +226,9 @@
 
+.text_preview-widget,
+.link_preview-widget {
+  .input-container{
```
oops
```suggestion
  .input-container {
```

zuk3975 commented on src/Adapter/Order/Repository/OrderRepository.php:
```diff
@@ -0,0 +1,73 @@
+     */
+    public function get(OrderId $orderId): Order
+    {
+        try {
+            $order = new Order($orderId->getValue());
+
+            if ($order->id !== $orderId->getValue()) {
+                throw new OrderNotFoundException($orderId, sprintf('%s #%d was not found', Order::class, $orderId->getValue()));
+            }
+        } catch (PrestaShopException $e) {
+            throw new CoreException(
+                sprintf(
+                    'Error occurred when trying to get %s #%d [%s]',
+                    Order::class,
+                    $orderId->getValue(),
+                    $e->getMessage()
+                ),
+                0,
+                $e
+            );
+        }
```
isn't that the same code as in AbstractObjectModelRepository::getObjectModel? You just provide an id, classname and exception class and the abstract method wraps it in try catch.

JevgenijVisockij replied:
Yeah it seems getObjectModel will do just fine, changed it.

JevgenijVisockij replied:
Actually had to change this back to the way it was. Can't use getObjectModel because OrderNotFound and OrderReturnStateNotFound exceptions need id in construct and so getObjectModel doesn't work.

zuk3975 commented on src/Adapter/OrderReturnState/Repository/OrderReturnStateRepository.php:
```diff
@@ -0,0 +1,71 @@
+     */
+    public function get(OrderReturnStateId $orderReturnStateId): OrderReturnState
+    {
+        try {
+            $orderReturnState = new OrderReturnState($orderReturnStateId->getValue());
+
+            if ((int) $orderReturnState->id !== $orderReturnStateId->getValue()) {
+                throw new OrderReturnStateNotFoundException($orderReturnStateId, sprintf('%s #%d was not found', OrderReturnState::class, $orderReturnStateId->getValue()));
+            }
+        } catch (PrestaShopException $e) {
+            throw new CoreException(
+                sprintf(
+                    'Error occurred when trying to get %s #%d [%s]',
+                    OrderReturnState::class,
+                    $orderReturnStateId->getValue(),
+                    $e->getMessage()
+                ),
+                0,
+                $e
+            );
+        }
```
Same as above

zuk3975 commented on src/Core/Domain/OrderReturn/QueryResult/OrderReturnForEditing.php:
```diff
@@ -0,0 +1,178 @@
+        string $customerFirstName,
+        string $customerLastName,
+        int $orderId,
+        DateTime $orderDate,
```
```suggestion
        DateTimeImmutable $orderDate,
```
wdyt?
There is also a DateTimeInterface which allows both DateTime and DateTimeImmutable, but I think in this case the date is only for view, so it shouln't ever be allowed to mutate.

JevgenijVisockij replied:
I think you are right, will add it as immutable.

zuk3975 commented on src/Core/Grid/Action/Row/Type/LinkRowAction.php:
```diff
@@ -68,6 +70,8 @@ protected function configureOptions(OptionsResolver $resolver)
             ->setAllowedTypes('accessibility_checker', [AccessibilityCheckerInterface::class, 'callable', 'null'])
             ->setAllowedTypes('clickable_row', 'boolean')
+            ->setAllowedTypes('disabled', 'boolean')
+
```
```suggestion
```

JevgenijVisockij committed:
Chganged repositores to use existing getObjectModel function

JevgenijVisockij committed:
Apply suggestions from code review

Some minor code review changes

Co-authored-by: Julius Žukauskas <31609858+zuk3975@users.noreply.github.com>

JevgenijVisockij committed:
Changed datetime as immutable

zuk3975 commented on src/Adapter/Order/Repository/OrderRepository.php:
```diff
@@ -29,43 +29,33 @@
 
 class OrderRepository extends AbstractObjectModelRepository
 {
+
```
```suggestion
```

JevgenijVisockij replied:
Done

JevgenijVisockij committed:
Fixed test to use immutable

JevgenijVisockij committed:
Removing uneeded imports

JevgenijVisockij committed:
Removing unneeded DateTime

jolelievre commented on src/Adapter/Order/Repository/OrderRepository.php:
```diff
@@ -0,0 +1,60 @@
+    public function get(OrderId $orderId): Order
+    {
+        /* @var Order $orderReturn */
+        $order = $this->getObjectModel($orderId->getValue(), Order::class, OrderNotFoundException::class);
```
This won't work because `OrderNotFoundException` expects an `OrderId` first parameter not a string message, I don't know if you noticed but your own tests are failing you should try and fix all the CIs before asking for review on a PR ^^

I'm think we could improve the `getObjectModel` so that it accepts a string for exception clas OR a callable to build it. So you could do:

```suggestion
        $order = $this->getObjectModel($orderId->getValue(), Order::class, function(int $id, string $objectModelClass, ?int $shopId) {
            return new OrderNotFoundException(new OrderId($id), sprintf('%s #%d was not found', $objectModelClass, $id));
        });
```

This requires a refacto of `AbstractModelRepository::fetchObjectModel` and change its parameters type, it's a breaking change but I think it's acceptable as long as the initial type (string) is still compatible.

```php
protected function fetchObjectModel(int $id, string $objectModelClass, string $exceptionClassOrCallback, ?int $shopId): ObjectModel
    {
        try {
            $objectModel = $this->constructObjectModel($id, $objectModelClass, $shopId);
            if ((int) $objectModel->id !== $id) {
                if (is_callable($exceptionClassOrCallback)) {
                    throw call_user_func($exceptionClassOrCallback, $id, $objectModelClass, $shopId);
                } else {
                    throw new $exceptionClassOrCallback(sprintf('%s #%d was not found', $objectModelClass, $id));
                }
            }
        } catch (PrestaShopException $e) {
            ...
        }

        return $objectModel;
    }
```

jolelievre replied:
This would be a great technical improvement for the whole repositories, if it works fine with this example we'll probably apply this to the other methods as well.

jolelievre commented on src/Adapter/Order/Repository/OrderRepository.php:
```diff
@@ -0,0 +1,60 @@
+        $order = $this->getObjectModel($orderId->getValue(), Order::class, OrderNotFoundException::class);
+        if (!($order instanceof Order)) {
+            throw new OrderException(sprintf('Must be instance of %s', Order::class));
+        }
```
This is overkill, the fact you're using the abstract repository makes it impossible to have another type. If you did this because of PHPStan know that the simple PHPDoc `/* @var Order $order */` is enough. In your case, it is miswritten that's why it doesn't work.

jolelievre replied:
```suggestion
```

jolelievre commented on src/Adapter/Order/Repository/OrderRepository.php:
```diff
@@ -0,0 +1,60 @@
+     */
+    public function get(OrderId $orderId): Order
+    {
+        /* @var Order $orderReturn */
```
```suggestion
        /* @var Order $order */
```

zuk3975 replied:
Jevgenij was obsessed with orderReturn for some reason in this PR :sweat_smile:

jolelievre commented on src/Adapter/OrderReturn/CommandHandler/UpdateOrderReturnStateHandler.php:
```diff
@@ -0,0 +1,101 @@
+        $orderReturn = $this->orderReturnRepository->get($orderReturnId);
+        $orderReturn = $this->updateOrderReturnWithCommandData($orderReturn, $command);
+
+        $this->orderReturnValidator->validate($orderReturn);
```
You should inject the validator inside your `orderReturnRepository` and call it directly from the `update` method. It will fully protect the update method no matter which code uses it, and it simplifies your DI as you don't need to inject both services each time

jolelievre replied:
You thumbed up, but it still needs to be done ^^ Or actually you handled it in the repository but still need to clean the code here no

JevgenijVisockij replied:
Yeah sorry I was thumbing up as I was adding changes, but I had not pushed commit yet. It was outdated because before it seems you did review at the same time i commited changes.

jolelievre replied:
Moving the validator is still needed here

jolelievre commented on src/Adapter/OrderReturn/CommandHandler/UpdateOrderReturnStateHandler.php:
```diff
@@ -0,0 +1,101 @@
+    public function handle(UpdateOrderReturnStateCommand $command): void
+    {
+        $orderReturnId = $command->getOrderReturnId();
+        $orderReturn = $this->orderReturnRepository->get($orderReturnId);
```
```suggestion
        $orderReturn = $this->orderReturnRepository->get($command->getOrderReturnId());
```

jolelievre replied:
Still needs to be simplified

jolelievre commented on src/Adapter/OrderReturn/QueryHandler/GetOrderReturnForEditingHandler.php:
```diff
@@ -0,0 +1,99 @@
+    public function handle(GetOrderReturnForEditing $query): OrderReturnForEditing
+    {
+        $orderReturnId = $query->getOrderReturnId();
+        $orderReturn = $this->orderReturnRepository->get($orderReturnId);
```
```suggestion
        $orderReturn = $this->orderReturnRepository->get($query->getOrderReturnId());
```

jolelievre commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -0,0 +1,87 @@
+        $orderReturn = $this->getObjectModel($orderReturnId->getValue(), OrderReturn::class, OrderReturnNotFoundException::class);
+        if (!($orderReturn instanceof OrderReturn)) {
+            throw new OrderReturnException(sprintf('Must be instance of %s', OrderReturn::class));
+        }
```
```suggestion
```
Here as well this should not be needed

jolelievre commented on src/Adapter/OrderReturnState/Repository/OrderReturnStateRepository.php:
```diff
@@ -0,0 +1,60 @@
+        $order = $this->getObjectModel($orderId->getValue(), OrderReturnState::class, OrderReturnStateNotFoundException::class);
+        if (!($order instanceof OrderReturnState)) {
+            throw new OrderReturnStateException(sprintf('Must be instance of %s', OrderReturnState::class));
+        }
```
```suggestion
```

jolelievre commented on src/Adapter/OrderReturnState/Repository/OrderReturnStateRepository.php:
```diff
@@ -0,0 +1,60 @@
+    public function get(OrderReturnStateId $orderId): OrderReturnState
+    {
+        /* @var OrderReturnState $orderReturn */
+        $order = $this->getObjectModel($orderId->getValue(), OrderReturnState::class, OrderReturnStateNotFoundException::class);
```
```suggestion
        /* @var OrderReturnState $orderReturnState */
        $orderReturnState = $this->getObjectModel($orderId->getValue(), OrderReturnState::class, OrderReturnStateNotFoundException::class);
```

jolelievre commented on src/Adapter/OrderReturnState/Repository/OrderReturnStateRepository.php:
```diff
@@ -0,0 +1,60 @@
+            throw new OrderReturnStateException(sprintf('Must be instance of %s', OrderReturnState::class));
+        }
+
+        return $order;
```
```suggestion
        return $orderReturnState;
```

jolelievre commented on src/PrestaShopBundle/Form/Admin/Type/LinkPreviewType.php:
```diff
@@ -0,0 +1,74 @@
+            'error_bubbling' => true,
+            'compound' => false,
+            'required' => false,
+            'button_label' => '', // Start automatically TinyMCE
```
```suggestion
            'button_label' => '',
```
I don't understand this comment 😅

jolelievre replied:
And why do you have all those new options? `button_label` alone is enough

JevgenijVisockij replied:
This preview type adds a button and button_label is what is written on button.

JevgenijVisockij committed:
Changing back not using getObjectModel function

jolelievre commented on src/PrestaShopBundle/Form/Admin/Type/LinkPreviewType.php:
```diff
@@ -0,0 +1,74 @@
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * This form type is used to display a text value without providing an interactive input to edit it.
```
This component is a good idea, however, you should update the description copied from `TextPreviewType` and explain the purpose of this new form type instead

jolelievre commented on src/Core/Form/IdentifiableObject/DataProvider/OrderReturnFormDataProvider.php:
```diff
@@ -0,0 +1,121 @@
+        return [
+            'question' => $orderReturnForEditing->getQuestion(),
+            'customer_name' => $orderReturnForEditing->getCustomerFullName(),
+            'customer_link' => $this->router->generate('admin_customers_view', [
```
it's original to handle links as form data but why not. The new form type to display links is indeed interesting

jolelievre commented on src/Core/Form/IdentifiableObject/DataProvider/OrderReturnFormDataProvider.php:
```diff
@@ -0,0 +1,121 @@
+            $orderReturnForEditing->getOrderId(),
+            $this->translator->trans('from', [], 'Admin.Global'),
+            $orderReturnForEditing->getOrderDate()->format($this->dateFormat)
+        );
```
I think you should choose a dedicated wording with placeholders instead:

```suggestion
        return $this->translator->trans(
            '#%order_id% from %order_date%',
            [
                '%order_id%' => $orderReturnForEditing->getOrderId(),
                '%order_date%' => $orderReturnForEditing->getOrderDate()->format($this->dateFormat),
            ],
            'Admin.Orderscustomers.Feature'
        );
```

jolelievre commented on src/Core/Grid/Action/Row/Type/LinkRowAction.php:
```diff
@@ -60,6 +61,7 @@ protected function configureOptions(OptionsResolver $resolver)
                 //route_param_name and route_param_field becomes redundant, but it cannot be removed due to BC break
                 'extra_route_params' => [],
                 'clickable_row' => false,
+                'disabled' => true,
```
Careful you can't suddenly add an option whose default behaviour is true and will impact all the existing use cases that didn't set this option to false. Besides I don't understand the purpose of this option which seems to only be used internally and is always forced to `true` ? Is it really needed? You don't even use it in the action you added

JevgenijVisockij replied:
Can't find reason for this change :/ I am guessing something went wrong during rebase.

jolelievre replied:
Seems like there is no modification in the `link.html.twig` associate template either so it's probably safer to revert this modification

jolelievre commented on src/PrestaShopBundle/Form/Admin/Sell/CustomerService/MerchandiseReturn/OrderReturnType.php:
```diff
@@ -0,0 +1,110 @@
+    {
+        $builder
+            ->add('customer_name', TextPreviewType::class, [
+                'default_empty_data' => self::CUSTOMER_NAME_PLACEHOLDER,
```
Why do you need those placeholders? You don't seem to perform any JS replacement or anything about this right? And the type is not used in a collection either so isn't this useless?

jolelievre commented on src/PrestaShopBundle/Form/Admin/Sell/CustomerService/MerchandiseReturn/OrderReturnType.php:
```diff
@@ -0,0 +1,110 @@
+            ])
+            ->add('question', TextPreviewType::class, [
+                'default_empty_data' => self::QUESTION_PLACEHOLDER,
+                'allow_html' => true,
```
Why do you need this new option? Besides it's not used in the template so it seems to be useless. And even if it weren't this could introduce some XSS injection I'm not sure it's a good idea, order message are supposed to be plain text.

JevgenijVisockij replied:
Merchandise return can be HTML and [Q&A reported bug](https://github.com/PrestaShop/PrestaShop/pull/24154#issuecomment-869607912) that customer explanation was not correctly displayed in merchandise return page, i see submiting merchandise return there is 
`            $orderReturn->question = htmlspecialchars(Tools::getValue('returnText'));
` So HTML is allowed from FO and I guess it makes sense it is displayed in BO as customer intended it.

jolelievre replied:
Ok that makes sense, I didn't see the modification in the theme previously All good

jolelievre commented on src/PrestaShopBundle/Form/Admin/Type/LinkPreviewType.php:
```diff
@@ -0,0 +1,74 @@
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults([
```
```suggestion
        parent::configureOptions($resolver);
        $resolver->setDefaults([
```

jolelievre replied:
Never forget to call the parent method, unless you have very good reasons to do so

jolelievre commented on src/PrestaShopBundle/Form/Admin/Type/TextPreviewType.php:
```diff
@@ -64,20 +65,27 @@ public function getBlockPrefix()
+                    'required' => false,
+                    'error_bubbling' => true,
+                    'compound' => false,
+                    'allow_html' => false, // Start automatically TinyMCE
```
Again with this weird comment 😅
And I don't understand the need of all those new options which are not even used?

JevgenijVisockij replied:
Yeah looked over and I don't think they are needed, they are the same in hidden type so should be succesfully inherited from it.

jolelievre commented on src/PrestaShopBundle/Resources/config/routing/admin/sell/customer_service/merchandise_return.yml:
```diff
@@ -23,3 +24,27 @@ admin_merchandise_returns_filter:
+    _controller: PrestaShopBundle\Controller\Admin\Sell\CustomerService\MerchandiseReturnController::editAction
+    _legacy_controller: AdminReturn
+    #    _legacy_link: AdminReturn:updateorder_return
+    #    _legacy_parameters:
```
I guess the `_legacy_parameters` can be left uncommented they don't enable the page anyway but it helps with automatic parameters reformatting

jolelievre reviewed (commented):
Thanks @JevgenijVisockij 

Lots of things done a good part of the feature is ready but I have some comments about needed improvement and suggestions for a cool refacto 😅

jolelievre reviewed (commented):
Hi @JevgenijVisockij 

I was reviewing this PR again, but I noticed most of my previous comments are not handled yet. So you should handle all of them and ping me hen it's ready.

Also, you need to rebase you branch this should fix both the conflicts and the PHPStan ci errors.

JevgenijVisockij committed:
Fixes from code review

JevgenijVisockij committed:
Update src/Core/Form/IdentifiableObject/DataProvider/OrderReturnFormDataProvider.php

Changed information whole block to be translated with variables

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

FabienPapet commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -0,0 +1,84 @@
+     */
+    public function update(OrderReturn $orderReturn): void
+    {
+        $this->orderReturnValidator->validate($orderReturn);
```
I'm not sure the update should validate.

JevgenijVisockij replied:
Well i checked some other repositories and they all do validate so I don't wanna break from current practice.
```
PrestaShop\PrestaShop\Adapter\Product\Repository\ProductRepository
PrestaShop\PrestaShop\Adapter\Customer\Repository\CustomerRepository
PrestaShop\PrestaShop\Adapter\Product\Combination\Repository\CombinationRepository
```

FabienPapet replied:
I don't think this is a good idea, because that is not the purpose of a repository. The `update` method should IMO act like a bridge between the object and the database.

If your repository can validate data, then it can also dispatch hooks, etc.. Then it is no more a repository, it's a manager.

JevgenijVisockij replied:
I see your point. Will change it.

zuk3975 replied:
Fabien is right about the principle, **but in this context** we put the **validation inside the repositories on purpose**. The reason is simple - those validators actually just wraps up the legacy obj model validation which will still happen inside the obj model later. So its not like we could choose to validate the object or not... By calling validators in repository, we just force the wrapping of legacy errors into new core exceptions with clear error codes and don't leave the opportunity for legacy exceptions to bubble up.

jolelievre replied:
I agreed with Fabien's vision initially (after a long discussion 😅) that a repository is not supposed to handle the validation or hook dispatch. But as Julius reminded me even if we don't validate it the ObjectModel handles validation internally anyway, except it won't throw the validation exception that we are no excepting. Meaning if anyone uses the repository without validating it first the handlers won't work as we expect.

So we absolutely need this validation to be done all the time and the safest and easiest way to do that is to include it in the repository so that it's always done. Besides it also simplifies the handlers because there is only one service to inject.

Now I agree the term Repository is probably not adapted in this case compared to the usual design pattern, but it's a bit late to rename all the services so I think we'll have to keep those for now. They are "Object Model Repositories" kind of a new custom thing from PrestaShop 😅

JevgenijVisockij committed:
Removed uneeded placeholders

JevgenijVisockij committed:
Adding change so allow html takes effect

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/Admin/Type/LinkPreviewType.php

Add parent method call

Co-authored-by: Jonathan Lelievre <jo.lelievre@gmail.com>

JevgenijVisockij committed:
Enabling legacy order id

JevgenijVisockij committed:
Adding interfaces for repositories

JevgenijVisockij committed:
Added interface for state repository changes handlers to use new repositories

JevgenijVisockij committed:
Removing uneeded line

JevgenijVisockij committed:
Uncommenting legacy parameter

JevgenijVisockij committed:
Fixing cs fixer and test

JevgenijVisockij committed:
CS fixer fix

FabienPapet commented on src/Core/Order/Repository/OrderRepositoryInterface.php:
```diff
@@ -0,0 +1,11 @@
+
+interface OrderRepositoryInterface
+{
+    public function get(OrderId $orderId): Order;
```
I think we can get rid of this legacy class in the core, we won't be able to remove it afterwards.
Order should have it's own interface also to make it compatible if possible. You can find an example here : https://github.com/Sylius/Order/blob/master/Model/OrderInterface.php

```suggestion
    public function get(OrderId $orderId);
```

JevgenijVisockij replied:
All the other repositories return legacy class as well, won't this mean that once we replace object model we will need to change them all anyway?

FabienPapet replied:
No, because if you are using an interface, you'll be able to construct a "new" object, with the methods. If you return legacy classes, we'll stick with public properties, magic getters, etc. We don't want to do that here.

jolelievre replied:
As I commented in another comment, this is not the moment to introduce such an interface, especially on the current OrderRepository which is meant to handle ObjectModel not generic interfaces

The work on generic core model interfaces and associated repositories will be done later but it needs some thought before, so you should remove this interface for now as it's not adapted

JevgenijVisockij committed:
Moving validate to state handler

JevgenijVisockij committed:
CS fixer fixes

FabienPapet commented on src/Adapter/OrderReturnState/Repository/OrderReturnStateRepository.php:
```diff
@@ -0,0 +1,74 @@
+     * @throws CoreException
+     * @throws OrderReturnStateNotFoundException
+     */
+    public function get(OrderReturnStateId $orderReturnStateId): OrderReturnState
```
Should return an object an `OrderReturnStateInterface` that implements all the getters , so we don't have legacy code in the new code.

jolelievre replied:
Same comment as previous repository, these repositories are not meant to return a core interface but ObjectModel instances

FabienPapet commented on src/Adapter/OrderReturn/Repository/OrderReturnRepository.php:
```diff
@@ -0,0 +1,75 @@
+     * @throws OrderReturnException
+     * @throws CoreException
+     */
+    public function get(OrderReturnId $orderReturnId): OrderReturn
```
Should return an object an `OrderReturnInterface` that implements all the getters , so we don't have legacy code in the new code.

zuk3975 replied:
Don't do it just yet. At least not in this PR. There are plenty of this kind repositories already and none of them yet have an interface for object model (also the abstract repository methods won't be compatable with the dedicated interfaces). These repositories were created with the Adapter namespace in mind and were expected to deal with the ObjectModel directly. (then the ADR of removing the Adapter was introduced).
If we want to start introducing interfaces for ObjectModels, then we need to think about it more. (think of what makes sense to have in such interface and what not, because current legacy ObjectModels has lots of crazy methods and I don't think we want to create contracts for all of them). If not thought through, this might lead into yet another mess.

jolelievre replied:
The plan here is not the handle the object model replacement, so the OrderReturnInterface, although it's a good idea, is not in this PR's scope we have other priorities for now

JevgenijVisockij replied:
Does the same apply to Repository interfaces or should I keep them?

jolelievre replied:
We discussed this with @FabienPapet for now these object model repositories shouldn't rely on new interfaces, their purpose is to help getting ObjectModel instances and they do that fine

If we want to introduce a new core OrderReturnInterface it should be done in a dedicated core repository which interface won't be related to ObjectModel (and maybe its implementation will rely on ObjectModel or this repo under the hood). But for now no need to change this service.

FabienPapet commented on src/Core/Form/IdentifiableObject/DataHandler/OrderReturnFormDataHandler.php:
```diff
@@ -0,0 +1,72 @@
+     */
+    public function create(array $data): void
+    {
+        @trigger_error(OrderReturn::class . ' is not created by form, this method should never be called', E_USER_WARNING);
```
throw exception here

JevgenijVisockij replied:
Why would exception be better? Exceptions are better in general because they can be caught and this error should never be caught because create method should never be used in the first place.

jolelievre replied:
I kind of agree with Fabien on this one, the method is never called as you said so anyway the exception should never be thrown. But if someone wrongly uses the form handler he needs to know right away that something is wrong, the `trigger_error` is a bit too silent that's why I think it's better to have an explicit exception that breaks everything so there's no ambiguity that something is not working.

We had many issues in the past because the legacy codes often fail silently, thus you don't know which step of a complex workflow has failed and need to investigate step by step to understand what went wrong. Since then I'm in favor of failing loudly and early 😅

I think we should add a `NotImplementedException extends CoreException` for those cases (there is one in the bundle namespace but we'd better use one from the core).

jolelievre commented on src/Core/OrderReturn/Repository/OrderReturnRepositoryInterface.php:
```diff
@@ -0,0 +1,13 @@
+use OrderReturn;
+use PrestaShop\PrestaShop\Core\Domain\OrderReturn\ValueObject\OrderReturnId;
+
+interface OrderReturnRepositoryInterface
```
Same here, you don't need this interface for now The repository doesn't need a core interface for now, and you don't need to use the interface in the handler either For now they wouldn't be correctly typed so it's counter productive

jolelievre commented on src/Core/OrderReturnState/Repository/OrderReturnStateRepositoryInterface.php:
```diff
@@ -0,0 +1,11 @@
+use OrderReturnState;
+use PrestaShop\PrestaShop\Core\Domain\OrderReturnState\ValueObject\OrderReturnStateId;
+
+interface OrderReturnStateRepositoryInterface
```
Same comment this should be removed

jolelievre reviewed (commented):
Thanks @JevgenijVisockij 

Just a few comments, mostly the validator still needs to be moved and the interfaces are too soon to introduce so you should remove them all.

JevgenijVisockij committed:
Removed interfaces

JevgenijVisockij committed:
Updated the form to use preview form types

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Type/TextPreviewType.php
#	src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml
#	templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Type/TextPreviewType.php

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Type/TextPreviewType.php

JevgenijVisockij committed:
Refactored repositories to work properly added integration tests for them.

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Type/TextPreviewType.php
#	src/PrestaShopBundle/Resources/config/services/adapter/order_return_state.yml

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Type/TextPreviewType.php

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Type/TextPreviewType.php

JevgenijVisockij committed:
Properly adding missing repositories and twig changes.

# Conflicts:
#	admin-dev/themes/new-theme/scss/components/_form.scss

JevgenijVisockij committed:
Yasml and css fixes

# Conflicts:
#	admin-dev/themes/new-theme/scss/components/_form.scss

JevgenijVisockij committed:
Removing uneeded types

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Type/TextPreviewType.php

JevgenijVisockij committed:
Moved validation back to repository

zuk3975 commented:
build failing :smile:

JevgenijVisockij committed:
Cs fixer fix

JevgenijVisockij committed:
EmployeeController.php

# Conflicts:
#	tests-legacy/PrestaShopBundle/Utils/DatabaseCreator.php
#	tests/Integration/Behaviour/behat.yml

# Conflicts:
#	tests/Integration/Behaviour/behat.yml

# Conflicts:
#	tests/Integration/Behaviour/behat.yml

# Conflicts:
#	tests/Integration/Behaviour/behat.yml

JevgenijVisockij committed:
CS fixer and stand fixes

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Type/TextPreviewType.php

# Conflicts:
#	src/Core/Grid/Action/Row/Type/LinkRowAction.php

JevgenijVisockij committed:
Fixed and improved behat test a bit

# Conflicts:
#	tests/Integration/Behaviour/behat.yml

JevgenijVisockij committed:
Removing not needed disabled changes

# Conflicts:
#	src/Core/Grid/Action/Row/Type/LinkRowAction.php

JevgenijVisockij committed:
Fixed tests for merchandise return

jolelievre reviewed (commented):
Thanks @JevgenijVisockij 

Only two things left to fix for me:
- revert the `disabled` attribute in LinkRowAction
- throw a `NotImplementedException` in the form handler

JevgenijVisockij committed:
Removing disabled from LinkRowAction.php

JevgenijVisockij committed:
Changed OrderReturnFormDataHandler.php to throw NotImplementedException.php

jolelievre reviewed (dismissed):
Thanks @JevgenijVisockij

JevgenijVisockij committed:
Fixing translate interface

jolelievre reviewed (dismissed):
Thanks @JevgenijVisockij all good

JevgenijVisockij committed:
Fixed proper interface

jolelievre reviewed (approved):
Thanks @JevgenijVisockij

jolelievre commented:
This is a partial PR the feature is not fully finished so we merge it early, the functional QA will be done on the last PR
